Annotations: Use a single row to represent a region

ryantxu opened the pull request:
Fixes: #17647, #15764, #12547

This PR adds a column for `epochEnd` to the annotation table and removes the need for two rows to represent a region.

Some things to note:
* backend uses `epoch` and `epochEnd`, frontend uses `time` and `timeEnd`
* this PR does not drop the `region_id` column, but just stops using it.  SQLite does not support dropping columns, you have to make a new table and copy into it.  I vote we leave the region_id column for now and fix it later... as alerts get more attention, I am sure they will want/need their own table.

On the frontend, `isRegion` is now defined with `time !== timeEnd`, we could try to update the API so that `timeEnd` only sends a value when it is not the same as time?

My motivation for this is to simplify the annotations structure before trying to port annotations to react in #17763 and to use any query result to represent annotations via something like #17280

ryantxu committed:
SQLite migrations

ryantxu committed:
cleanup

ryantxu committed:
migrate end times

marefr commented on pkg/services/sqlstore/migrations/annotation_mig.go:
```diff
@@ -109,4 +109,24 @@ func addAnnotationMig(mg *Migrator) {
+		Cols: []string{"epoch_end"}, Type: IndexType,
+	}))
+	mg.AddMigration("Make epoch_end the same as epoch", NewRawSqlMigration("UPDATE annotation SET epoch_end = epoch"))
+	mg.AddMigration("Move existing regions to epoch_end", NewRawSqlMigration(
```
MySQL integration test are failing and are usually not that great at supporting advanced update clauses. I think we need to use a code migration here instead, see example here:
https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/sqlstore/migrations/user_mig.go#L118
https://github.com/grafana/grafana/blob/219d711597fb34e53aba065190311af02c887200/pkg/services/sqlstore/migrations/user_mig.go#L127-L155

Probably need to do it in batches for those Grafana instances having a huge amount of annotations.

marefr commented:
Looks like a great start :+1: Really impressive you can do backend as well. Had no idea :smile:

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  Elasticsearch: Visualize logs in Explore (#17605)
  Grafana-CLI: Wrapper for `grafana-cli` within RPM/DEB packages and config/homepath are now global flags (#17695)
  Add guidelines for SQL date comparisons (#17732)
  Docs: clarified usage of go get and go mod (#17637)
  Project: Issue triage doc improvement (#17709)
  Improvement: Grafana release process minor improvements (#17661)
  TimePicker: New time picker dropdown & custom range UI (#16811)
  RemoteCache: redis connection string parsing test (#17702)
  Fix link in pkg/README (#17714)
  Dashboard: Use Explore's Prometheus editor in dashboard panel edit (#15364)
  Settings: Fix typo in defaults.ini (#17707)
  Project: Adds a security policy (#17698)
  Project: Adds support resource docs (#17699)
  Document issue triage process (#17669)
  noImplicitAny: slate (#17681)
  config: fix connstr for remote_cache (#17675)
  Explore: Improves performance of Logs element by limiting re-rendering (#17685)
  Docs: Flag serve_from_sub_path as available in 6.3 (#17674)

ryantxu committed:
switch to update with a query

ryantxu committed:
real migration

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  TimePicker: fixed minor issues with new timepicker (#17756)
  Explore: Parses and updates TimeSrv in one place in Explore (#17677)
  @grafana/ui: release (#17754)
  Password: Remove PasswordStrength (#17750)
  Devenv:SAML: devenv block with saml test app (#17733)
  LDAP:Docs: add information on LDAP sync feature and update LDAP sync default (#17689)
  Graph: Add data links feature (click on graph) (#17267)
  Explore: Changes LogsContainer from a PureComponent to a Component (#17741)
  Chore: Remove tether and tether drop dependency in grafana/ui (#17745)
  noImplicitAny: time region manager etc. (#17729)
  Panel: Fully escape html in drilldown links (was only sanitized before)  (#17731)
  Alerting: Improve alert rule testing (#16286)

ryantxu committed:
anno migrations

ryantxu committed:
remove old docs

torkelo reviewed (commented):
This is looking really good! looks to simplify regions a lot

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  release: update latest.json to v6.2.5 (#17767)
  release: 6.2.5 changelog (#17766)
  Fix typo s/Applicaiton/Application/ in error messages (#17765)
  UserAdmin: UI for disabling users (#17333)
  API: get list of users with additional auth info (#17305)

ryantxu committed:
set isRegion from time changes

ryantxu committed:
use <> for is not

ryantxu commented:
> you can do backend as well. Had no idea 😄

Mile wide, inch deep!

marefr commented on packages/grafana-ui/src/types/data.ts:
```diff
@@ -104,4 +104,5 @@ export interface AnnotationEvent {
   text?: string;
   type?: string;
   tags?: string[];
+  source?: any; // source.type === 'dashboard' ????
```
Hmm where is this used or what is it used for?

ryantxu replied:
ya, that one is weird.  In the existing code, the sources gets added on each query and then checked in:
https://github.com/grafana/grafana/blob/master/public/app/features/annotations/annotations_srv.ts#L46

I'll add a comment in the code.  I think we should rethink this (but not in this PR)

marefr commented on pkg/api/annotations.go:
```diff
@@ -69,43 +68,26 @@ func PostAnnotation(c *m.ReqContext, cmd dtos.PostAnnotationsCmd) Response {
+		tmp := item.Epoch
+		item.EpochEnd = item.Epoch
+		item.Epoch = tmp
+	}
```
Guess it's okay to switch times around. Alternative would be to return an error instead.

ryantxu replied:
I think the API should swap -- frontend can then drag left or right and not worry

marefr replied:
Make sense

marefr commented on pkg/api/annotations.go:
```diff
@@ -247,23 +216,21 @@ func PatchAnnotation(c *m.ReqContext, cmd dtos.PatchAnnotationsCmd) Response {
+		tmp := existing.Epoch
+		existing.EpochEnd = existing.Epoch
+		existing.Epoch = tmp
+	}
```
Same comment as for creating annotation.

marefr commented on pkg/services/sqlstore/annotation.go:
```diff
@@ -67,10 +70,18 @@ func (r *SqlAnnotationRepo) Update(item *annotations.Item) error {
+		if existing.EpochEnd == 0 {
+			existing.EpochEnd = item.Epoch
+		}
+		if existing.EpochEnd > existing.Epoch {
```
Hmm, here you have the same logic again as in the API for create/patch annotation. Could this be put in one go package, preferrably only in sqlstore package/this file?

ryantxu replied:
good idea.  I removed it from the api, and only have it in sqlstore.  There is now a function that is called for both Save and Update

marefr replied:
Great

marefr commented on pkg/services/sqlstore/migrations/annotation_mig.go:
```diff
@@ -109,4 +110,51 @@ func addAnnotationMig(mg *Migrator) {
+	// 6.3: Make Regions a single annotation row
+	//
+	mg.AddMigration("Add epoch_end column", NewAddColumnMigration(table, &Column{
+		Name: "epoch_end", Type: DB_BigInt, Nullable: false, Default: "0",
```
Think this needs to be nullable true to handle existing data is this table. Integration tests probably works because there are no data in the table, but I need to verify this by manual testing.

ryantxu replied:
I tested with SQLite and it works... but yes, we need to manually test other databases migrations.  May be good to have our tests start with something and try to migrate... but beyond my skills now :)

marefr commented on pkg/services/sqlstore/migrations/annotation_mig.go:
```diff
@@ -109,4 +110,51 @@ func addAnnotationMig(mg *Migrator) {
+		Name: "epoch_end", Type: DB_BigInt, Nullable: false, Default: "0",
+	}))
+	mg.AddMigration("Add index for epoch_end", NewAddIndexMigration(table, &Index{
+		Cols: []string{"epoch_end"}, Type: IndexType,
```
Think this index isn't useful since that would only be in use when filter annotations only by `epoch_end`. What we probably need is an index combined of org_id, epoch and epoch_end. 

Adding an index is maybe something that we can wait with to add when we see that it's actually needed?

ryantxu replied:
I just changed it to:
```
	mg.AddMigration("Add index for epoch_end", NewAddIndexMigration(table, &Index{
		Cols: []string{"org_id", "epoch", "epoch_end"}, Type: IndexType,
	}))
```
But we should for sure check if this actually gets used in the real query!

marefr commented on pkg/services/sqlstore/migrations/annotation_mig.go:
```diff
@@ -109,4 +110,51 @@ func addAnnotationMig(mg *Migrator) {
+}
+
+func (m *AddMakeRegionSingleRowMigration) Exec(sess *xorm.Session, mg *Migrator) error {
+	regions := make([]*TempRegionInfoDTO, 0)
```
Maybe we should consider using a transaction? Not sure what possible consequences that may have...

ryantxu replied:
I thought each migration already happens in a transaction?  See:

https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/migrator/migrator.go#L97

```
		err := mg.inTransaction(func(sess *xorm.Session) error {
			err := mg.exec(m, sess)
                         ...
```

marefr replied:
Yeah that's true :+1:

marefr reviewed (changes requested):
Great progress :+1: Added a couple of comments, but feels like we should be able to have this in a merge:able state soon.

Something to consider is the annotation migration we did for v5.2, see [upgrade notes](https://grafana.com/docs/installation/upgrading/#upgrading-to-v5-2). When we know which release this may land in we probably need to add similar upgrade notes.

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  Usage Stats: Update known datasource plugins (#17787)
  Docs: Adds section on Querying Logs for Elasticsearch (#17730)
  Docs: Adds section on Querying Logs for InfluxDB (#17726)
  Devenv: makes the grafana users default for saml. (#17782)
  Explore: Displays only one Time column as configured in TimeZone settings (#17775)
  Markdown: Replace rendering library (#17686)
  ApiKeys: Fix check for UTC timezone (#17776)
  Prometheus: Minor style fix (#17773)
  Docs: fixed notifications table
  Auth: Allow expiration of API keys (#17678)
  17278 prometheus step align utc (#17477)
  Docs: Update release guide  (#17759)

ryantxu committed:
add comment and fix index decleration

ryantxu committed:
single validation place

ryantxu committed:
add test

ryantxu committed:
fix test

ryantxu committed:
add upgrading docs

ryantxu commented:
Thanks for the review @marefr -- super helpful

I think I have addressed all your comments the remaining TODOs are:
1. decide if this goes in 6.3  (I hope so)
2. manual testing for postgresql etc migrations
3. maybe investigate if the indicies for `epoch` and `epochEnd` are used in the find query

ryantxu committed:
move to 6.4

ryantxu committed:
use AnnotationEvent

marefr commented:
@ryantxu Don't think that this PR fixes #15010. Typo?

Kimberlite commented:
Just allow me to use your conversation to ask you a question : it is possible to create an annotation range in elasticsearch and display it in grafana ? a priori, the only possibility is to keyboard the annotation range in grafana at the opposite of the simple annotation marker who can be created in elasticsearch... In other words, can you use annotation range in scripted grafana ? Many thanks for your reply

torkelo commented:
After this is merged it should be an easy fix

ryantxu commented:
> Don't think that this PR fixes #15010. Typo?
@marefr -- dooh, good catch.  I keep getting tricked by githubs autocomplete when I paste an issue number.  you have to hit escape or it will autocomplete to something else!

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master: (22 commits)
  grafana/toolkit: bundle plugins with webpack (#17850)
  Explore: Adds orgId to URL for sharing purposes (#17895)
  grafana/toolkit: copy sass files (#17888)
  ChangePassword: Rewrite change password page to react (#17811)
  AngularPanels: Fixed loading state indication for angular panels (#17900)
  Explore: Adds support for toggling text edit mode in explore (#17870)
  LDAP: Divide the requests (#17885)
  Build: fixes missing shebang in release tagging script. (#17894)
  Teams: show proper label for each auth provider (#17860)
  Logging: Login and Logout logging actions (#17760) (#17883)
  Loki: Adds comment explaining usage of RFC3339Nano string (#17872)
  Explore: Query rows are now reset when changing data sources (#17865)
  Codestyle: add guidelines for removing the m alias for models (#17890)
  Docs: How to work with themes (#17876)
  Docs: Fix developing plugins index page (#17877)
  StatsPicker: Fix multiple value input layout etc. (#17827)
  Chore: Build grafana-cli when running bra run (#17788)
  Build: use golangci-lint as a make command (#17739)
  Explore: Log highlights only update when user stops typing (#17845)
  Loki: getHighlighterExpressionsFromQuery Returns null if filter term is not quoted (#17852)
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master: (27 commits)
  @grafana/data: export dateMath and rangeUtil (#17971)
  Refactor: move some files to @grafana/data (#17952)
  noImplicitAny: Azure Monitor (#17966)
  grafana/toolkit: initial CI task and various small improvements (#17914)
  First version of prettier checks in toolkit (#17964)
  LDAP: finishing touches (#17945)
  Graphite: Refactor lexer and parser (#17958)
  noImplicitAny: Datasource files, under 2500 (#17959)
  Auth: saml enabled check. (#17960)
  Auth: SAML login button. (#17932)
  grafana/toolkit: Add support for extensions styling (#17938)
  Datasource: Refactor Graphite to class (#17942)
  SAML: Configuration defaults, examples and dependencies  (#17954)
  OAuth: deny login for disabled users (#17957)
  Build: Adds pre-commit check that fails if node versions are not synced (#17820)
  Docs: minor ha-setup edit (#17950)
  Docs: Added very basic docs about revoking user sessions (#17931)
  Docs: Updates documentation regarding logs integration in Explore (#17896)
  noImplicitAnys: Fix InfluxDB type issues #17937)
  TimePicker: align position between dashboard and explore (#17940)
  ...

ryantxu committed:
fix import

torkelo commented:
Tested this on POSTGRES but something wrong with the migration, the regions after the migration are not regions any more (timeEnd same as time when fetched from api)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master: (46 commits)
  Docs: Simplify download links & instructions and make download link clearer (#18090)
  FieldDisplay: move threshold and mapping to Field (#17043)
  InfluxDB: Enable interpolation within ad-hoc filter values for InfluxDB data source (#18077)
  Docs: Move data links down (#18072)
  grafana/toolkit: improve CircleCI integration (#18071)
  Build: consistently reference go binary (#18059)
  devenv: Fix typo in nginix docker for mac (#18068)
  noImplicitAny: 1670 errors (#18035)
  Add missing pull requests to Changelog (#18061)
  provisioning: escape literal '$' with '$$' to avoid interpolation (#18045)
  grafana/toolkit: improve CircleCI stubs (#17995)
  Docs: clarify the ttl units (#18039)
  Update docs readme for running MySQL/Postgres tests
  Auth: Duplicate API Key Name Handle With Useful HTTP Code (#17905)
  Chore: upgrade node-sass to 4.12.0 (#18052)
  API: Minor fix for nil pointer when trying to log error during creating new dashboard via the API (#18003)
  Chore: update lodash (#18055)
  Update latest.json (#18043)
  Update Changelog (#18042)
  Chore: bump master version number to 6.4.0-pre
  ...

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  ValuMapping: start with some values (#18092)

torkelo commented:
We should try to get this to mergable state, could not get migration to work (regions turned to single events when tested with postgres)

ryantxu commented:
What's the best way to test postgres?  Any tips?

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master: (82 commits)
  TablePanel: Remove scroll option on TablePanel (#18318)
  Keyboard Shortcuts: Sign in to enable them (#18271)
  GitHub Templates: Pull Request Template update (#18300)
  Auth Proxy: Include additional headers as part of the cache key (#18298)
  grafana/toolkit: support windows paths (#18306)
  Chore: noImplicitAny Sub 500 errors (#18287)
  Plugins: return a promise for loadPluginCss (#18273)
  Utils: avoid calling console.warn() too often for deprecation warnings (#18269)
  CLI: Allow installing custom binary plugins (#17551)
  Docs: Update link to example app (#18253)
  GettingStarted: Skip Query for getting started (#18268)
  v6.3.0-beta2 is latest testing (#18283)
  Release: Changelog update with v6.3.0-beta2 (#18281)
  Chore: Upgrades typescript to version 3.5 (#18263)
  docs: team sync (#18239)
  SAML: Only show SAML login button on Enterprise version (#18270)
  Permissions: Show plugins in nav for non admin users but hide plugin configuration (#18234)
  CI: Change target branch in CI task trigger-docs-update (#18255)
  Plugins: Include build number and PR in metadata (#18260)
  Run End-to-End tests for release builds (#18211)
  ...

torkelo commented:
I used the devenv Postgres block

torkelo commented:
./devenv/create_docker_compose.sh postgres
docker-compose up


```ini
[database]
# You can configure the database connection by specifying type, host, name, user and password
# as separate properties or as on string using the url properties.

# Either "mysql", "postgres" or "sqlite3", it's your choice
; type = postgres
; host = localhost:5432
; name = grafana
; user = grafana
; 
```

marefr commented:
Please refer to the following instructions: https://github.com/grafana/grafana#running-the-mysql-or-postgres-backend-tests

ryantxu committed:
merge master

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master: (24 commits)
  Explore: Fix loading error for empty queries (#18488)
  Fix: Fixes stripping of $d in Explore urls (#18480)
  grafana/ui: fix toTimeTicks error (#18448)
  Docs: Adds details to Pull Request Checklist (#18471)
  DataLinks: respect timezone when displaying datapoint's timestamp in graph context menu (#18461)
  Chore: strictNullChecks, ColoringEditor and time_region_manager (#18442)
  Backend: Do not set SameSite cookie attribute if cookie_samesite is none (#18462)
  DataLinks: Apply scoped variables correctly (#18454)
  DataLinks: Use datapoint timestamp correctly when interpolating variables (#18459)
  API: Minor fix for team creation endpoint when using API key (#18252)
  Login: Adjust space between skip and its icon (#18407)
  Docs: Update Auth Proxy documentation (#18444)
  Docs: Minor Readme update (#18438)
  OAuth: return GitLab groups as a part of user info (enable team sync) (#18388)
  Fix: Avoid glob of single-value array variables  (#18420)
  DataLinks: Enable multiple data links per panel (#18434)
  Markdown: Handle undefined/null strings (#18433)
  Update CHANGELOG.md
  Update CHANGELOG.md
  Docs: Update changelog and latest with 6.3.1 and 6.3.2 releases (#18437)
  ...

ryantxu committed:
remove regionId from typescript

ryantxu commented:
@torkelo -- I just tested with postgres and it seems to have worked ok.

Before patch:
![image](https://user-images.githubusercontent.com/705951/62811444-80d88b00-bab6-11e9-9427-c0431bb40737.png)

Startup:
![image](https://user-images.githubusercontent.com/705951/62811515-d6ad3300-bab6-11e9-97a4-3fe5055ced2c.png)

After:
![image](https://user-images.githubusercontent.com/705951/62811459-9057d400-bab6-11e9-9bfc-73b1fdf4b204.png)

Note that:
1. epoch_end colum exists and is filled in
2. two rows were removed (10,12) and the end times are set in the new rows: 9,10

Regions look fine:
![image](https://user-images.githubusercontent.com/705951/62811549-ef1d4d80-bab6-11e9-8daf-024798a8c492.png)


Can you tell me more about your environment?

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  Fix: failing build after React bump (#18514)
  strictNullChecks: First batch (#18390)
  Chore: bump React to 16.9.0 (#18502)
  Docs: Adds a new security section (#18508)
  Docs: Update issue triage doc with external PRs section (#18464)
  Typo: fix typo in processDataFrame.ts comment (#18492)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  Login: Angular to React (#18116)
  InfoTooltip: Info icon with tooltip (#18478)
  Annotations: Fix failing annotation query when time series query is cancelled (#18532)
  remotecache: support SSL with redis (#18511)
  QueryData: Handle that response data must be array (#18504)
  React: Rename deprecated UNSAFE_componentWillReceiveProps (#18526)
  Explore: Replaces TimeSeries with GraphSeriesXY (#18475)
  API: Restrict anonymous user information access (#18422)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into anno-regions

* grafana/master:
  DataFrame: convert from row based to a columnar value format (#18391)
  Packages: Temporarily skip canary releases if packages build fail (#18577)
  Update latest.json to latest stable version (#18575)
  Docs: Update changelog for v6.3.3 (#18569)
  Graph: Fixed issue clicking on series line icon (#18563)
  grafana/toolkit: Unpublish previous "next" version when releasing a new one (#18552)
  Toolkit: write PR report to a folder with the circle build number (#18560)
  CI: Fail build if yarn.lock is not up to date (#18555)
  Chore: Updates react-dependant packages to address react warnings (#18549)
  Prometheus: Fix regression of rerunning query on legend/interval change (#18147)
  Explore/Prometheus: More consistently allows for multi-line queries (#18362)
  Login: Fixes undefined redirect (#18545)
  Plugins: expose react-redux, redux (#18501)
  TimeSeries: assume values are all numbers (#18540)

ryantxu commented:
@torkelo --- ping...  I think it is OK.  if not, I need more info about how you tested it.

torkelo commented:
Ok, tested again and cannot replicate my issue. Sorry, I think the frontend part must not have been rebuilt after I switched to this branch so the backend migration ran but without updated frontend, all region events showed as single events.

fangel commented on public/app/features/annotations/annotations_srv.ts:
```diff
@@ -60,7 +60,10 @@ export class AnnotationsSrv {
-        annotations = makeRegions(annotations, options);
+        for (let i = 0; i < annotations.length; i++) {
+          const a = annotations[i];
+          a.isRegion = a.time !== a.timeEnd;
```
I know I'm a little late to the party, but I was working on #17602, and wanted to adopt it to this new method of creating regions.
All the existing single-point annotations from Elasticsearch is rather broken right now, because the ES annotations only set `time`, not `timeEnd`, and thus they all get turned into regions.

Wouldn't it be more backwards compatible to do `a.isRegion = a.timeEnd && a.time !== a.timeEnd` or something similar.
That way the data-sources that don't provide a `timeEnd` doesn't see their annotations turned into regions?

marefr replied:
Think #18903 fixes this. Please merge in/rebase on master

fangel replied:
Ah, right - I hadn't seen that it had been fixed later - sorry. So yes, I'll rebase #17602 and re-test.
